src/chainstate/stacks/index/storage.rs

```diff
@@ -0,0 +1,1304 @@
+            self.dump_traverse(f, &root, &hash, parent_bhh)
+        }
+        else {
+            trace!("Failed to dump {:?}: not the current block", bhh);
+            Err(Error::NotFoundError)
+        }
+    }
+
+    fn size_hint(&self) -> usize {
+        self.total_bytes
+    }
+}
+
+impl TrieStorage for TrieRAM {
+    fn extend(&mut self, bhh: &BlockHeaderHash) -> Result<(), Error> {
+        if self.block_header == *bhh {
+            return Err(Error::ExistsError);
+        }
+        trace!("Extend to {:?}", bhh);
+        self.block_header = bhh.clone();
+        self.offset = 0;
+        self.num_nodes = 0;
+        self.data.clear();
+        self.readonly = false;
+        Ok(())
+    }
+
+    fn open(&mut self, bhh: &BlockHeaderHash, readwrite: bool) -> Result<(), Error> {
+        if self.block_header != *bhh {
+            trace!("Failed to open {:?}: not the current block", bhh);
+            return Err(Error::NotFoundError);
+        }
+        self.block_header = bhh.clone();
+        self.offset = 0;
+        self.num_nodes = self.data.len() as u64;
+        self.readonly = !readwrite;
+        Ok(())
+    }
+
+    fn tell(&self) -> BlockHeaderHash {
+        self.block_header.clone()
+    }
+    
+    fn block_walk(&mut self, _back_block: u32) -> Result<BlockHeaderHash, Error> {
+        panic!("Not implemented for TrieRAM");
+    }
+
+    fn root_ptr(&self) -> u64 { 0 }
+
+    fn readwrite(&self) -> bool {
+        !self.readonly
+    }
+
+    fn format(&mut self) -> Result<(), Error> {
+        if self.readonly {
+            trace!("Read-only!");
+            return Err(Error::ReadOnlyError);
+        }
+
+        self.data.clear();
+        self.offset = 0;
+        self.num_nodes = 0;
+        Ok(())
+    }
+
+    fn read_node_hash_bytes(&mut self, ptr: &TriePtr, buf: &mut Vec<u8>) -> Result<(), Error> {
+        if (ptr.ptr() as u64) >= (self.data.len() as u64) {
+            trace!("TrieRAM: Failed to read node bytes: {} >= {}", ptr.ptr(), self.data.len());
+            Err(Error::NotFoundError)
+        }
+        else {
+            fast_extend_from_slice(buf, self.data[ptr.ptr() as usize].1.as_bytes());
+            Ok(())
+        }
+    }
+
+    fn read_node(&mut self, ptr: &TriePtr) -> Result<(TrieNodeType, TrieHash), Error> {
+        let disk_ptr = ftell(self)?;
+        trace!("TrieRAM: read_node({:?}): at {}: {:?}", &self.block_header, disk_ptr, ptr);
+
+        self.read_count += 1;
+        if is_backptr(ptr.id()) {
+            self.read_backptr_count += 1;
+        }
+        else if ptr.id() == TrieNodeID::Leaf {
+            self.read_leaf_count += 1;
+        }
+        else {
+            self.read_node_count += 1;
+        }
+
+        if (ptr.ptr() as u64) >= (self.data.len() as u64) {
+            trace!("TrieRAM: Failed to read node: {} >= {}", ptr.ptr(), self.data.len());
+            Err(Error::NotFoundError)
+        }
+        else {
+            Ok(self.data[ptr.ptr() as usize].clone())
```

Comment:
it seems like there's a range mismatch between the storage and the Trie pointers -- `self.data` can have up to 2^64 entries, while `ptr()` is a `u32` -- what happens if the Trie grows to be larger than 4GB?

Reply:
It won't grow bigger than 4GB, since a single block won't produce that much data (each block will produce its own `Trie`, and each `Trie` will got into its own file).

The reason for the mismatch is because `TrieRAM` implements `Seek` which necessarily means 64-bit offsets.  However, `TriePtr`s only support 32-bit pointers, since a `Trie` will never be that big and we could use the extra space.